File: portfolio-backend/Controllers/VisitorsController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using portfolio_backend.Data;
using portfolio_backend.Models;
using System;
using System.Linq;
using System.Threading.Tasks;

namespace portfolio_backend.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class VisitorsController : ControllerBase
    {
        private readonly ApplicationDbContext _context;

        public VisitorsController(ApplicationDbContext context)
        {
            _context = context;
        }

        // POST: api/visitors/track
        [HttpPost("track")]
        public async Task<IActionResult> TrackVisit()
        {
            try {
                // Get IP address from request
                var ipAddress = HttpContext.Connection.RemoteIpAddress?.ToString() ?? "unknown";
                var userAgent = Request.Headers["User-Agent"].ToString();

                // Simple check to avoid double-counting same IP within a 24-hour window
                var lastVisit = await _context.Visitors
                    .Where(v => v.IpAddress == ipAddress)
                    .OrderByDescending(v => v.Timestamp)
                    .FirstOrDefaultAsync();

                if (lastVisit == null || (DateTime.UtcNow - lastVisit.Timestamp).TotalHours >= 24)
                {
                    var visitor = new Visitor
                    {
                        IpAddress = ipAddress,
                        UserAgent = userAgent.Length > 100 ? userAgent.Substring(0, 100) : userAgent,
                        Timestamp = DateTime.UtcNow
                    };

                    _context.Visitors.Add(visitor);
                    await _context.SaveChangesAsync();
                    return Ok(new { message = "Visit tracked successfully" });
                }

                return Ok(new { message = "Visit already recorded recently" });
            }
            catch (Exception ex) {
                return BadRequest(new { error = ex.Message });
            }
        }

        // GET: api/visitors/stats
        [HttpGet("stats")]
        public async Task<IActionResult> GetStats()
        {
            try {
                var totalVisitors = await _context.Visitors.CountAsync();
                var todayVisitors = await _context.Visitors
                    .Where(v => v.Timestamp >= DateTime.UtcNow.Date)
                    .CountAsync();

                return Ok(new { 
                    TotalVisitors = totalVisitors,
                    TodayVisitors = todayVisitors
                });
            }
            catch (Exception ex) {
                return BadRequest(new { error = ex.Message });
            }
        }
    }
}

File: portfolio-backend-nestjs/src/visitors/visitors.controller.ts
import { Controller, Get, Post, Req, Ip, BadRequestException } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';

@Controller('api/visitors')
export class VisitorsController {
  constructor(private prisma: PrismaService) {}

  @Post('track')
  async trackVisit(@Req() req: any, @Ip() ipAddress: string) {
    try {
      const userAgent = req.headers['user-agent'] || '';
      const clientIp = ipAddress || 'unknown';

      // Check if visited in last 24h
      const lastVisit = await this.prisma.visitor.findFirst({
        where: { ipAddress: clientIp },
        orderBy: { timestamp: 'desc' },
      });

      const twentyFourHoursAgo = new Date();
      twentyFourHoursAgo.setHours(twentyFourHoursAgo.getHours() - 24);

      if (!lastVisit || lastVisit.timestamp < twentyFourHoursAgo) {
        await this.prisma.visitor.create({
          data: {
            ipAddress: clientIp,
            userAgent: userAgent.length > 100 ? userAgent.substring(0, 100) : userAgent,
            timestamp: new Date(),
          },
        });
        return { message: 'Visit tracked successfully' };
      }

      return { message: 'Visit already recorded recently' };
    } catch (ex: any) {
      throw new BadRequestException({ error: ex.message });
    }
  }

  @Get('stats')
  async getStats() {
    try {
      const totalVisitors = await this.prisma.visitor.count();

      // Today visitors (since midnight UTC or midnight local? EF Core uses DateTime.UtcNow.Date which is midnight UTC)
      const today = new Date();
      today.setUTCHours(0, 0, 0, 0);

      const todayVisitors = await this.prisma.visitor.count({
        where: {
          timestamp: {
            gte: today,
          },
        },
      });

      return {
        totalVisitors,
        todayVisitors,
      };
    } catch (ex: any) {
      throw new BadRequestException({ error: ex.message });
    }
  }
}
